### PR TITLE
Clean up ProtocolMessage and use Either not Try

### PR DESCRIPTION
**Pull request by kirkwood**
This is a clean-up pull request.

The `ProtocolMessage` object no longer does anything but handle protocol buffer construction and deserialization into Scala types (that wrap protocol buffer types). As mentioned in #24 in response to @DoSOfRedRiver. Functionality specific to each protocol message class (`PingMessage`, _etc._) has been moved into that class. Common functions (`header` and serialization) are elevated to the `ProtocolMessage` trait.

This PR also has a first pass at removing all the `Try` types from this level. One remains in the interface to the Kademlia table (the return value from the `ping` callback). I may have gone a little overboard with the types, but I want to put this out there before I keep at it.

I haven't yet looked at SAM and how that might help me with the receiving thread in `UnicastNetwork` (which I still haven't really addressed, since the caller isn't written, yet).

### CODE REVIEW

File: comm/src/main/scala/coop/rchain/comm/unicast.scala [hunk 1]
```diff
@@ -76,13 +76,15 @@ case class UnicastComm(local: PeerNode) extends Comm {
     * exception if the underlying socket receive
     * throws. See @java.net.DatagramSocket.receive.
     */
-  override def recv: Try[Seq[Byte]] =
-    Try {
+  override def recv: Either[CommError, Seq[Byte]] =
```
Comment (tymm) [inline comm/src/main/scala/coop/rchain/comm/unicast.scala]: Please also update the comment :)

File: comm/src/main/scala/coop/rchain/comm/unicast.scala [hunk 2]
```diff
@@ -92,13 +94,18 @@ case class UnicastComm(local: PeerNode) extends Comm {
     * the exception in a @scala.util.Try. See
     * @java.net.DatagramSocket.send for a list of possible exceptions.
     */
-  override def send(data: Seq[Byte], peer: PeerNode): Try[Unit] =
+  override def send(data: Seq[Byte], peer: PeerNode): Either[CommError, Unit] =
```
Comment (tymm) [inline comm/src/main/scala/coop/rchain/comm/unicast.scala]: Update comment.

File: comm/src/main/scala/coop/rchain/comm/unicast.scala [hunk 3]
```diff
@@ -1,12 +1,13 @@
...
 import java.net.{DatagramPacket, DatagramSocket}
-import scala.util.{Failure, Success, Try}
+import scala.util.control.NonFatal
 
-sealed trait DatagramError extends Throwable
+sealed trait DatagramError extends CommError
 case class DatagramOverflow(size: Int) extends DatagramError
 case class DatagramUnderflow(size: Int) extends DatagramError
-case class DatagramFramingError(size: Int) extends DatagramError
+case class DatagramFramingError(ex: Throwable) extends DatagramError
```
Comment (tymm) [inline comm/src/main/scala/coop/rchain/comm/unicast.scala]: See previous comment about `Throwable`.

File: comm/src/main/scala/coop/rchain/comm/unicast.scala [hunk 4]
```diff
@@ -1,12 +1,13 @@
...
+case class DatagramException(ex: Throwable) extends DatagramError
```
Comment (tymm) [inline comm/src/main/scala/coop/rchain/comm/unicast.scala]: See previous comment about `Throwable`.

File: comm/src/main/scala/coop/rchain/comm/comm.scala [hunk 5]
```diff
@@ -1,10 +1,11 @@
 package coop.rchain.comm
 
-import scala.util.Try
+trait CommError
```
Comment (DoSOfRedRiver) [inline comm/src/main/scala/coop/rchain/comm/comm.scala]: Please use `sealed trait`, this prevents someone to extend your ADT and helps compiler to check that your pattern matching expressions is exhaustive.

File: comm/src/main/scala/coop/rchain/comm/network.scala [hunk 6]
```diff
@@ -25,19 +24,20 @@ case class UnicastNetwork(id: NodeIdentifier, endpoint: Endpoint) extends Protoc
...
       while (true) {
         comm.recv match {
-          case Success(res) =>
+          case Right(res) =>
             for {
               msg <- ProtocolMessage.parse(res)
             } dispatch(msg)
-          case Failure(ex: SocketTimeoutException) => ()
-          case Failure(ex)                         => ex.printStackTrace
+          case Left(ex: SocketTimeoutException) => ()
```
Comment (DoSOfRedRiver) [inline comm/src/main/scala/coop/rchain/comm/network.scala]: It seems like you should pattern match against `CommError` here and below.

File: comm/src/main/scala/coop/rchain/comm/protocol.scala [hunk 7]
```diff
@@ -5,6 +5,10 @@ import coop.rchain.comm.protocol.routing._
 import scala.util.{Failure, Success, Try}
 import scala.concurrent.duration.{Duration, MILLISECONDS}
 
+sealed trait ProtocolError extends CommError
```
Comment (DoSOfRedRiver) [inline comm/src/main/scala/coop/rchain/comm/protocol.scala]: It is a good practice to define all elements of an ADT in one place, so we could use `sealed` modificator and easily observe all elements in one place.

File: comm/src/main/scala/coop/rchain/comm/protocol.scala [hunk 8]
```diff
@@ -48,20 +52,21 @@ class ProtocolNode(id: NodeIdentifier, endpoint: Endpoint, handler: ProtocolHand
...
+      case Right(resp) =>
+        req.header match {
+          case Some(incoming) =>
+            Success(Duration(resp.timestamp - incoming.timestamp, MILLISECONDS))
+          case _ => Failure(new Exception("ping failed"))
         }
-      case None => Failure(new Exception("ping failed"))
+      case Left(ex) => ex match {
+        case ProtocolException(exc) => Failure(exc)
+        case exc => Failure(new Exception(exc.toString))
```
Comment (DoSOfRedRiver) [inline comm/src/main/scala/coop/rchain/comm/protocol.scala]: I'd prefer to use `Exception` as constructor argument, so we could have a full stack trace.

File: comm/src/main/scala/coop/rchain/comm/unicast.scala [hunk 9]
```diff
@@ -61,44 +62,48 @@ case class UnicastComm(local: PeerNode) extends Comm {
...
     */
-  override def recv: Try[Seq[Byte]] =
-    Try {
+  override def recv: Either[CommError, Seq[Byte]] =
+    try {
       receiver.receive(recv_dgram)
       decode(recv_dgram.getData) match {
-        case Right(bytes) => bytes
-        case Left(err)    => throw(err)
+        case Right(bytes) => Right(bytes)
```
Comment (DoSOfRedRiver) [inline comm/src/main/scala/coop/rchain/comm/unicast.scala]: It seems like we can remove this `match` expression.

File: comm/src/main/scala/coop/rchain/comm/unicast.scala [hunk 10]
```diff
@@ -61,44 +62,48 @@ case class UnicastComm(local: PeerNode) extends Comm {
...
-        Try(sender.send(dgram))
+        try {
+          sender.send(dgram)
+          Right(())
+        } catch {
+          case NonFatal(ex) => Left(DatagramException(ex))
+        }
       }
-      case Left(err) => Failure(err)
+      case Left(err) => Left(err)
```
Comment (DoSOfRedRiver) [inline comm/src/main/scala/coop/rchain/comm/unicast.scala]: Also, if you don't use left branch, you can just `flatMap` over `encode(data)`.

File: comm/src/main/scala/coop/rchain/comm/network.scala [hunk 11]
```diff
@@ -121,25 +121,27 @@ case class UnicastNetwork(id: NodeIdentifier, endpoint: Endpoint) extends Protoc
...
         val pend = PendingKey(remote.key, header.timestamp, header.seq)
-        val promise = Promise[Try[ProtocolMessage]]
+        val promise = Promise[Either[ProtocolError, ProtocolMessage]]
         pending.put(pend, promise)
         try {
           comm.send(bytes, remote)
           Await.result(promise.future, timeout)
         } catch {
-          case ex: Throwable => Failure(ex)
+          case ex: Throwable => Left(ProtocolException(ex))
```
Comment (DoSOfRedRiver) [inline comm/src/main/scala/coop/rchain/comm/network.scala]: Not sure, are we going to replace `Throwable` here with `Exception`?

File: comm/src/main/scala/coop/rchain/comm/unicast.scala [hunk 12]
```diff
@@ -57,48 +52,45 @@ case class UnicastComm(local: PeerNode) extends Comm {
...
-    * This method returns a @scala.util.Try, which may contain an
-    * exception if the underlying socket receive
-    * throws. See @java.net.DatagramSocket.receive.
+    * Returns `Right` with the bytes read from the socket or Left with
+    * an error, if something went wrong.
     */
-  override def recv: Try[Seq[Byte]] =
-    Try {
+  override def recv: Either[CommError, Seq[Byte]] =
+    try {
```
Comment (DoSOfRedRiver) [inline comm/src/main/scala/coop/rchain/comm/unicast.scala]: If you wish you can still use `Try` here:
```scala
Try {
  receiver.receive(recv_dgram)
  decode(recv_dgram.getData)
} recover {
  case NonFatal(ex: Exception) => Left(DatagramException(ex)) 
}
```

**Pull request merged**